go/vt/vttablet/grpctmclient/cached_client.go

```diff
@@ -0,0 +1,322 @@
+	if err != nil {
+		dialer.connWaitSema.Release()
+		return nil, nil, err
+	}
+
+	cc, err := grpcclient.DialContext(ctx, addr, grpcclient.FailFast(false), opt)
+	if err != nil {
+		dialer.connWaitSema.Release()
+		return nil, nil, err
+	}
+
+	dialer.m.Lock()
+	defer dialer.m.Unlock()
+
+	if conn, existing := dialer.conns[addr]; existing {
+		// race condition: some other goroutine has dialed our tablet before we have;
+		// this is not great, but shouldn't happen often (if at all), so we're going to
+		// close this connection and reuse the existing one. by doing this, we can keep
+		// the actual Dial out of the global lock and significantly increase throughput
+		cc.Close()
+		dialer.connWaitSema.Release()
+		return dialer.redialLocked(conn)
+	}
+
+	conn := &cachedConn{
+		TabletManagerClient: tabletmanagerservicepb.NewTabletManagerClient(cc),
+		cc:                  cc,
+		lastAccessTime:      time.Now(),
+		refs:                1,
+		addr:                addr,
+	}
+	dialer.evict = append(dialer.evict, conn)
+	// dialer.evictSorted = false -- no need to do this here
+	dialer.conns[addr] = conn
+
+	return dialer.connWithCloser(conn)
+}
+
+// redialLocked takes an already-dialed connection in the cache does all the work of
+// lending that connection out to one more caller. this should only ever be
+// called while holding at least the RLock on dialer.m (but the write lock is
+// fine too), to prevent the connection from getting evicted out from under us.
+//
+// It returns the three-tuple of client-interface, closer, and error that the
+// main dial func returns.
+func (dialer *cachedConnDialer) redialLocked(conn *cachedConn) (tabletmanagerservicepb.TabletManagerClient, io.Closer, error) {
+	dialerStats.ConnReuse.Add(1)
+	conn.lastAccessTime = time.Now()
+	conn.refs++
+	dialer.evictSorted = false
+	return dialer.connWithCloser(conn)
+}
+
+// connWithCloser returns the three-tuple expected by the main dial func, where
+// the closer handles the correct state management for updating the conns place
+// in the eviction queue.
+func (dialer *cachedConnDialer) connWithCloser(conn *cachedConn) (tabletmanagerservicepb.TabletManagerClient, io.Closer, error) {
+	return conn, closeFunc(func() error {
+		dialer.m.Lock()
+		defer dialer.m.Unlock()
+		conn.refs--
+		dialer.evictSorted = false
+		return nil
+	}), nil
+}
+
+// Close closes all currently cached connections, ***regardless of whether
+// those connections are in use***. Calling Close therefore will fail any RPCs
+// using currently lent-out connections, and, furthermore, will invalidate the
+// io.Closer that was returned for that connection from dialer.dial().
+//
+// As a result, it is not safe to reuse a cachedConnDialer after calling Close,
+// and you should instead obtain a new one by calling either
+// tmclient.TabletManagerClient() with
+// TabletManagerProtocol set to "grpc-cached", or by calling
+// grpctmclient.NewCachedConnClient directly.
```

Comment:
Similarly, this comment is flat-out not true in the simpler implementation, so I'll clean that up as well

Reply:
Ah, I lied, this is partially true, since the closer func still locks the `dialer.m` it will still content with actual dials if you reuse a cachedConnDialer after close, but it won't actually mess with the state of the queue. So it's "safe" but not ideal.